meta/autoid/autoid.go

```diff
@@ -479,27 +579,52 @@ func (alloc *allocator) AllocSeqCache(tableID int64) (int64, int64, int64, error
 }
 
 func validIncrementAndOffset(increment, offset int64) bool {
+	failpoint.Inject("skipIncrementOffsetValidation", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(true)
+		}
+	})
 	return (increment >= minIncrement && increment <= maxIncrement) && (offset >= minIncrement && offset <= maxIncrement)
 }
 
-// CalcNeededBatchSize is used to calculate batch size for autoID allocation.
+// calcNeededBatchSize is used to calculate batch size for autoID allocation.
 // It firstly seeks to the first valid position based on increment and offset,
 // then plus the length remained, which could be (n-1) * increment.
-func CalcNeededBatchSize(base, n, increment, offset int64, isUnsigned bool) int64 {
+func (alloc *allocator) calcNeededBatchSize(base, n, increment, offset int64) (batchSize int64, expectedEnd int64, isOverflow bool) {
 	if increment == 1 {
-		return n
+		overflow, expectedEnd := alloc.plus(base, n)
+		if overflow {
+			return n, 0, true
+		}
+		return n, expectedEnd, false
+	}
+	overflow, diff := alloc.multiply(n-1, increment)
+	if overflow {
+		return 0, 0, true
 	}
-	if isUnsigned {
-		// SeekToFirstAutoIDUnSigned seeks to the next unsigned valid position.
-		nr := SeekToFirstAutoIDUnSigned(uint64(base), uint64(increment), uint64(offset))
-		// Calculate the total batch size needed.
-		nr += (uint64(n) - 1) * uint64(increment)
-		return int64(nr - uint64(base))
+	firstID := alloc.seekToFirstAutoID(base, increment, offset)
+	overflow, expectedEnd = alloc.plus(firstID, diff)
+	if overflow {
+		return 0, 0, true
 	}
-	nr := SeekToFirstAutoIDSigned(base, increment, offset)
-	// Calculate the total batch size needed.
-	nr += (n - 1) * increment
-	return nr - base
+	return expectedEnd - base, expectedEnd, false
+}
+
+func (alloc *allocator) adjustStep(startTime time.Time) {
+	if alloc.customStep {
+		return
+	}
+	consumeDur := startTime.Sub(alloc.lastAllocTime)
+	alloc.step = NextStep(alloc.step, consumeDur)
+}
+
+func (alloc *allocator) calcActualStep(base, batchSize int64) (actualStep int64) {
+	actualStep = alloc.step
+	if isOverflow, _ := alloc.plus(base, actualStep); isOverflow {
+		// base < batchSize < alloc.step
```

Comment:
why batchSize < alloc.step always？

Reply:
Because `base + batchSize` doesn't overflow but `base + alloc.step` does.
The overflow check of `base+batchSize` has been done in `calcNeededBatchSize` already.